### PR TITLE
RFC for vector length agnostic SVE Vectorized class

### PR DESCRIPTION
**Pull request by Ryo-not-rio**
This is a proposal for a Vectorized class for SVE to replace the current Vectorized class which will make it vector length agnostic, avoiding code duplication in the future

### CODE REVIEW

File: RFC-0044-sve-vectorized-class.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+## **Drawbacks**
+### Implementation cost
+This is a large change which requires an overhaul of all of the current SVE Vectorized as well as any code that expects the size() function to be constexpr. The first cost can be mitigated by updating the Vectorized classes one by one, but the size() change will need to be done all at once.
+
+### Sideffects from non-constexpr size()
+There are a number of functions that use the size() function to initialize an array. These will have to be changed to an alternative such as a vector. Since a vector is implemented as an array under the hood, we hope this will not cause any regressions but a thorough benchmarking of these functions need to be done to ensure that this is the case.
+
+## **Alternatives**
+To keep the size() function constexpr, we considered setting the size of the Vectorized class to be the maximum possible SVE vector length (currently 512 bits) and loading multiple vectors as necessary. However, this poses the following problems:
```
Comment (jondea) [inline RFC-0044-sve-vectorized-class.md]: Might be worth clarifying that 2048 is the maximum possible, 512 is the maximum available hardware

File: RFC-0044-sve-vectorized-class.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+Now this introduces quite an obvious overhead of an additional load and store operation with each op. However, the compiler is able to optimize these out with the following conditions:
+
+1. The -O3 flag is set
+2. The `svptrue_b32()` predicate is used
+3. You are storing to and then loading from the same pointer
+
+Ensuring these conditions are met and by inlining the functions, we can rely on the compiler to optimize the duplicate load and stores, ensuring we do not introduce any regressions.
+
+### The size problem
+We face a challenge with this implementation due to the constraint of the size() function being constexpr. The size() function which returns the number of elements in the Vectorized class cannot be constexpr in our implmentation due to SVE vector lengths being unknown at compile time. We propose we change this to be const instead of constexpr.
```
Comment (jondea) [inline RFC-0044-sve-vectorized-class.md]: It may be worth noting that I think a const virtual function can be overridden by a constexpr function. And even if not, I suspect in many cases the compiler should be able to work it out if the class is concrete

File: RFC-0044-sve-vectorized-class.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+## **Proposed Implementation**
+The basic premise of our proposal is to store not the vector but an array in our vectorized class which we will load from and store to with each operation. A minimal version is shown at the end.
```
Comment (jondea) [inline RFC-0044-sve-vectorized-class.md]: I assume here you don't mean std::vector, might be worth clarifying.

File: RFC-0044-sve-vectorized-class.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+3. You are storing to and then loading from the same pointer
+
+Ensuring these conditions are met and by inlining the functions, we can rely on the compiler to optimize the duplicate load and stores, ensuring we do not introduce any regressions.
+
+### The size problem
+We face a challenge with this implementation due to the constraint of the size() function being constexpr. The size() function which returns the number of elements in the Vectorized class cannot be constexpr in our implmentation due to SVE vector lengths being unknown at compile time. We propose we change this to be const instead of constexpr.
+
+```
+class Vectorized<float> {
+	float values[64]; // Maximum number of elements supported by any SVE machine
```
Comment (jondea) [inline RFC-0044-sve-vectorized-class.md:29]: What's the expected additional memory/allocation overhead of having the max? I assume that there are relatively few of the these classes around and they are not created/destroyed in hot loops, so it should be minimal.

File: RFC-0044-sve-vectorized-class.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+## **Motivation**
+PyTorch contains a "Vectorized" class that wraps different SIMD intrinsics for different architectures by having a vector as a class attribute, and the class methods acting as the wrapper around the intrinsics.  This works well for non-scalable vectors but poses an issue for SVE due to the inability to store them as class attributes. The current workaround for this is to use the compiler flag -msve-vector-bits=\<bits\>, however this is not ideal as this would 1. require separate vectorized classes for different vector lengths and 2. does not allow for runtime detection of the actual vector length. We currently only have an implementation of the Vectorized class for 256-bit SVE machines but as we think about adding support for different vector length, we need to consider how to avoid code duplication as raised by @malfet [here](https://github.com/pytorch/pytorch/pull/138388#issuecomment-2635612409). This RFC aims to solve the issue but creating a Vectorized class that detects the vector length at runtime as SVE is intended to be used, allowing us to support different vector lengths without writing any duplicate code.
```
Comment (jondea) [inline RFC-0044-sve-vectorized-class.md]: nit: you refer to the `Vectorized` class in 4 different ways (backticks, speechmarks, caps etc)

**Comment by jondea** [comment]
I left a few minor comments, but I think overall it looks promising! Thank you

**Review by albanD — COMMENTED** [review]
cc @malfet @swolchok @kimishpatel @guangy10 as people working on this topic in case you have any comments

**Comment by kimishpatel** [comment]
> cc @malfet @swolchok @kimishpatel @guangy10 as people working on this topic in case you have any comments

thanks Alban. Will take a look tomorrow or friday

File: RFC-0044-sve-vectorized-class.md (hunk 2 above)
Comment (swolchok) [reply RFC-0044-sve-vectorized-class.md]: I don't think there are virtual functions involved; architectures specialize Vectorized<T>

File: RFC-0044-sve-vectorized-class.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  - X * sigmoid * softmax sped up by 1.65x on Neoverse V2
+- No speed or accuracy regression on 256-bit vectors
+
+
+## **Drawbacks**
+### Implementation cost
+This is a large change which requires an overhaul of all of the current SVE Vectorized as well as any code that expects the size() function to be constexpr. The first cost can be mitigated by updating the Vectorized classes one by one, but the size() change will need to be done all at once.
+
+### Sideffects from non-constexpr size()
+There are a number of functions that use the size() function to initialize an array. These will have to be changed to an alternative such as a vector. Since a vector is implemented as an array under the hood, we hope this will not cause any regressions but a thorough benchmarking of these functions need to be done to ensure that this is the case.
```
Comment (swolchok) [inline RFC-0044-sve-vectorized-class.md]: I would be concerned about this. Arrays go on the stack; vectors go on the heap.

File: RFC-0044-sve-vectorized-class.md (hunk 1 above)
Comment (swolchok) [reply RFC-0044-sve-vectorized-class.md]: Is "size() determines the size of an array" the only use of constexpr size()? If so, we should have a few options to solve that problem:
- rely on the VLA extension (I'm mobile so I can't look up which compilers do and don't support this, but I'd expect clang and gcc to get it right and msvc to be annoying) to size arrays this way anyway; we would just have to use C arrays instead of std::array
- use some template metaprogramming magic to create a container templated on vectorized that detects whether size is constexpr. Create an array if so, and either a padded-to-max-length-for-sve array or a vector otherwise.

Comment (swolchok) [reply RFC-0044-sve-vectorized-class.md]: Some quick research says we can generally detect whether a function is constexpr when we get C++20, which should be soon: https://quuxplusone.github.io/blog/2022/01/04/test-constexpr-friendliness/

However, since this is a very specific case, we could just create a custom vectorized_of_t_has_constexpr_size traits class that defaults to true and is specialized for VLA-SVE-specialized Vectorized<T> types to say false.

Comment (Ryo-not-rio) [reply RFC-0044-sve-vectorized-class.md]: Another option to keep size() constexpr is to use a hybrid approach where we decide the size() on compile time. This will keep size() constexpr but will not be able to take advantage of SVE's runtime vector length detection feature

Comment (Ryo-not-rio) [reply RFC-0044-sve-vectorized-class.md:71]: After further investigation, changing size() to const would require a rewrite of the VectorizedN class as well so I'm leaning towards the hybrid approach mentioned above

File: RFC-0044-sve-vectorized-class.md (hunk 4 above)
Comment (Ryo-not-rio) [reply RFC-0044-sve-vectorized-class.md:29]: Added information on this!

File: RFC-0044-sve-vectorized-class.md (hunk 3 above)
Comment (Ryo-not-rio) [reply RFC-0044-sve-vectorized-class.md]: fixed

File: RFC-0044-sve-vectorized-class.md (hunk 5 above)
Comment (Ryo-not-rio) [reply RFC-0044-sve-vectorized-class.md]: fixed

File: RFC-0044-sve-vectorized-class.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+## **Proposed Implementation**
+The basic premise of our proposal is to store not the SVE vector but an array in our `Vectorized` class which we will load from and store to with each operation. A minimal version is shown at the end.
+
+Now this introduces quite an obvious overhead of an additional load and store operation with each op. However, the compiler is able to optimize these out with the following conditions:
+
+1. The -O3 flag is set
+2. The `svptrue_b32()` predicate is used
+3. You are storing to and then loading from the same pointer
+
+Ensuring these conditions are met and by inlining the functions, we can rely on the compiler to optimize the duplicate load and stores, ensuring we do not introduce any regressions.
```
Comment (cfRod) [inline RFC-0044-sve-vectorized-class.md:22]: Is there any risk of different behaviour across different compilers/versions?

File: RFC-0044-sve-vectorized-class.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+     	svst1_f32(svptrue_b32(), values, v);
+		return *this;
+  	}
+}
+```
+
+## **Metrics **
+- Reduction of code duplication
+- Speedup of PyTorch on SVE machines with non-256 bit vectors
+  - Softmax sped up by 2.73x on Neoverse V2
```
Comment (cfRod) [inline RFC-0044-sve-vectorized-class.md:58]: Are these numbers similar for eager mode or compile?

File: RFC-0044-sve-vectorized-class.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+## **Alternatives**
+To keep the size() function constexpr, we considered setting the size of the `Vectorized` class to be the maximum possible SVE vector length and loading multiple vectors as necessary. However, this poses the following problems:
+
+1. Increases the tail loop size unecessarily for machines with smaller vector lengths.
+2. Compiler can no longer optimize out duplicate loads and stores as multiple vectors need to be handled consecutively.
+3. Detection of number of vectors needed introduces an extra overhead
+
+Due to these issues combined, especially 2., this alternative introduces a ~30% overhead compared to the current implementation.
+
+## **Unresolved questions**
```
Comment (cfRod) [inline RFC-0044-sve-vectorized-class.md:93]: Are there any plans to add specific tests?

Comment (Ryo-not-rio) [reply RFC-0044-sve-vectorized-class.md:93]: For unit tests, the existing tests should cover us, for perf tests, we should run the whole set of integration tests

File: RFC-0044-sve-vectorized-class.md (hunk 8 above)
Comment (Ryo-not-rio) [reply RFC-0044-sve-vectorized-class.md:58]: For eager mode

**Comment by Ryo-not-rio** [comment]
@swolchok I've scoped out the all the changes that need to be done to change `size()` to not const-expr and added a benchmark plan to make these changes, please have a look and let me know what you think

File: RFC-0044-sve-vectorized-class.md (hunk 7 above)
Comment (Ryo-not-rio) [reply RFC-0044-sve-vectorized-class.md:22]: Yes, I guess we'll have to make sure this optimization happens on all our supported compilers

File: RFC-0044-sve-vectorized-class.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+## **Proposed Implementation**
+The basic premise of our proposal is to store not the SVE vector but an array in our `Vectorized` class which we will load from and store to with each operation. A minimal version is shown at the end.
+
+Now this introduces quite an obvious overhead of an additional load and store operation with each op. However, the compiler is able to optimize these out with the following conditions:
+
+1. The -O3 flag is set
```
Comment (kimishpatel) [inline RFC-0044-sve-vectorized-class.md:18]: -O3 can introduce other issues, including code bloat. how do you handle that?

File: RFC-0044-sve-vectorized-class.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+     	svst1_f32(svptrue_b32(), values, v);
+		return *this;
+  	}
+}
+```
+
+## **Metrics **
+- Reduction of code duplication
+- Speedup of PyTorch on SVE machines with non-256 bit vectors
+  - Softmax sped up by 2.73x on Neoverse V2
```
Comment (kimishpatel) [inline RFC-0044-sve-vectorized-class.md:58]: I think one question i have is this: true power of SVE I presume is that you can have same binary compiled for one implementation of sve work on another and realize performance gain. That is great. If this means that pytorch distribution built for arm, when pip installed, can take advantage of vector length of that machine. is that how we plan to realize performance gain?

File: RFC-0044-sve-vectorized-class.md (hunk 10 above)
Comment (Ryo-not-rio) [reply RFC-0044-sve-vectorized-class.md:18]: If I'm not mistaken, PyTorch is built with -O3 in release mode so there shouldn't be any extra code bloat

File: RFC-0044-sve-vectorized-class.md (hunk 11 above)
Comment (Ryo-not-rio) [reply RFC-0044-sve-vectorized-class.md:58]: Yes, currently, we only have SVE support for 256bit vector machines, with this change we can do as you suggested: build one version of PyTorch for aarch64 and use it on all aarch64 machines without worrying about vector length

File: RFC-0044-sve-vectorized-class.md (hunk 10 above)
Comment (kimishpatel) [reply RFC-0044-sve-vectorized-class.md:18]: That for pypi release, but deployment of pytorch may get built with different flags. Plus note that this being header only change also allows reuse of the class in other implementations including that of custom ops that dont live inside pytorch. And my custom op lib may not be built with O3 for size reasons. Imagine for example deploying this on mobile. so I do think we should touch upon this aspect carefully and at least call it out clearly.

**Comment by swolchok** [comment]
> @swolchok I've scoped out the all the changes that need to be done to change `size()` to not const-expr and added a benchmark plan to make these changes, please have a look and let me know what you think

this doesn't seem to address my concerns about adding additional heap allocations. are there technical barriers to the plan I outlined to avoid heap allocations, such as reasons for size() to be constexpr other than determining the size of an array? We don't need benchmarking to tell us that heap allocations are worse than no heap allocations.

**Comment by Ryo-not-rio** [comment]
>are there technical barriers to the plan I outlined to avoid heap allocations, such as reasons for size() to be constexpr other than determining the size of an array

So I've investigated all the places where size() needs to be constexpr [here](https://github.com/Ryo-not-rio/pytorch/commit/fa05c1de3340215da5dc0a32612e75e2816fc143) and as you suggest, a lot of it just initializing an array which can be solved by using c arrays. The other places where it's used are template parameters where we would have to move the parameter from the template to a function argument

File: RFC-0044-sve-vectorized-class.md (hunk 10 above)
Comment (Ryo-not-rio) [reply RFC-0044-sve-vectorized-class.md:18]: Thanks for your insight, that is something I hadn't considered so I'll put it into the RFC. I'm not sure if there's any mitigation we could do at the moment, I'll try less aggressive optimization flags to see how much the compiler can optimize this aspect

**Comment by malfet** [comment]
I probably commented on PyTorch issue rather than RFC, so repeating here this [comment](https://github.com/pytorch/pytorch/issues/153471#issuecomment-2880442090):
> constexpr size allows for a numerous optimizations on the compiler side, and making it dynamic will likely result in a significant slowdown. To accept change like that, that will affect all supported CPU architectures a rigorous test plan is necessary as well as list of HW one must run it against.

**Comment by Ryo-not-rio** [comment]
Since we're adding SVE128 support, I think we can pause this for now, and revisit in the future if needed
